### PR TITLE
[ENH] refactoring of babs_submit function and updator of job status table

### PR DESCRIPTION
**Pull request by djarecka**
@zhao-cy - I've started also working on the option of adding job arrays (it's important for our group), and I've created a new function to remove some duplication in `babs_submit`, especially because with the job array option I would add even more duplication.
I've decided to create a separate PR for this, so this doesn't mix with the array changes

### CODE REVIEW

**Comment by zhao-cy** [comment]
Oh this is a great idea! Thank you @djarecka !

**Comment by djarecka** [comment]
@zhao-cy - I've also created a function `prepare_job_ind_list` that prepare list of job indices that should be submitted. I think this simplify `babs_submit`, but let me know what you think. 

Today, I only tested this for single session

File: babs/utils.py [hunk 1]
```diff
@@ -1394,6 +1394,131 @@ def submit_one_job(analysis_path, type_session, type_system, sub, ses=None,
...
+        dataframe of jobs to be submitted (specified by user)
+    count: int
+        number of jobs to be submitted
+    type_session: str
+        type of session, can be "single-ses" or "multi-ses"
+
+    Returns:
+    ------------------
+    job_ind_list: list
+        list of job indices to be submitted
```
Comment (zhao-cy) [inline babs/utils.py:1463]: Would add further explanation like "Here, job index refers to index in the full job status dataframe `df_job`."

File: babs/babs.py [hunk 2]
```diff
@@ -1012,119 +1014,32 @@ def babs_submit(self, count=1, df_job_specified=None):
...
 
-        j_count = 0
-
         try:
             with lock.acquire(timeout=5):  # lock the file, i.e., lock job status df
                 df_job = read_job_status_csv(self.job_status_path_abs)
+                # getting list of jobs indices to submit (based either on df_job_specified or count):
+                job_ind_list = prepare_job_ind_list(df_job, df_job_specified, count, self.type_session)
                 df_job_updated = df_job.copy()
+                for i_progress, i_job in enumerate(job_ind_list):
```
Comment (zhao-cy) [inline babs/babs.py:1023]: Before this, not sure if you need to add a sanity check of "if `job_ind_list` is empty"? i.e., when there is no job to submit.

**Comment by zhao-cy** [comment]
@djarecka I think adding `prepare_job_ind_list()` function is smart! I added two minor comments on the code changes - see above. Let me know if you cannot see my comments! Thank you Dorota!!

File: babs/babs.py (hunk 2 above)
Comment (djarecka) [reply babs/babs.py:1023]: i think it should be fine, it will just do nothing

**Comment by djarecka** [comment]
I've also run a couple test for for multi-session!

**Comment by zhao-cy** [comment]
@djarecka Wonderful! I believe you have tested out different `babs-submit` arguments (like `--job`, `--count`) and `babs-status` arguments (like `--resubmit`) under different scenarios (e.g., some jobs are pending/running/failed)? I'm also happy to help testing out if needed! Just let me know!

And once these testings are done let me know if you think this PR is ready go? If so I'll review your code and if everything looks good to me I'll merge it to main.

**Comment by djarecka** [comment]
I've tried to test for various scenarios for `babs-submit` (with `--job` and  `--count`) and `babs-status`, but I haven't tested `--resubmit`...

Perhaps it would be great if you can also run couple tests, just in case?

Otherwise, this is ready, for `array` I prefer to open a separate pr

**Comment by zhao-cy** [comment]
Sure sounds good! Will run some tests, review your code and merge to main!

**Comment by zhao-cy** [comment]
List of things I've tested using a multi-session dataset:
- [ ] `babs-submit`
  - [ ] `--job` x 2 (i.e., for two sessions), and checked the `job_status.csv` along the way. Jobs were successfully finished 
- [ ] `babs-status`
  - [ ] basic use, i.e., `--project-root` argument only
  - [ ] `resubmit failed`, checked `job_status.csv` after resubmitting
  - [ ] `resubmit-job <failed sub and ses>`, checked `job_status.csv` after resubmitting

**Review by zhao-cy — COMMENTED** [review]
Hi @djarecka! I've reviewed the code changes. I have several comments. Most of my comments are just suggestions of making comments in the code more clear, and only one of my comment is for a minor bug.

In addition I've tested this PR on Slurm cluster - I listed what I've tested above. I originally planned to test on our SGE cluster, however it was down today... But since the changes are not related to cluster system type, I think it's okay to skip testing on SGE cluster. Let me know if you think differently.

**Comment by djarecka** [comment]
sorry I've included some changes by mistake

**Pull request merged**